vortex-array/src/array/primitive/compute/compare.rs

```diff
@@ -1,24 +1,48 @@
 use arrow_buffer::bit_util::ceil;
-use arrow_buffer::{BooleanBuffer, MutableBuffer};
+use arrow_buffer::{BooleanBuffer, BooleanBufferBuilder, MutableBuffer};
 use vortex_dtype::{match_each_native_ptype, NativePType};
-use vortex_error::VortexResult;
+use vortex_error::{VortexExpect, VortexResult};
+use vortex_scalar::PrimitiveScalar;
 
 use crate::array::primitive::PrimitiveArray;
-use crate::array::BoolArray;
-use crate::compute::{CompareFn, Operator};
+use crate::array::{BoolArray, ConstantArray};
+use crate::compute::{MaybeCompareFn, Operator};
+use crate::validity::ArrayValidity;
 use crate::{Array, IntoArray, IntoArrayVariant};
 
-impl CompareFn for PrimitiveArray {
-    fn compare(&self, other: &Array, operator: Operator) -> VortexResult<Array> {
-        let other = other.clone().into_primitive()?;
-
-        let match_mask = match_each_native_ptype!(self.ptype(), |$T| {
-            apply_predicate(self.maybe_null_slice::<$T>(), other.maybe_null_slice::<$T>(), operator.to_fn::<$T>())
-        });
-
-        let validity = self.validity().and(other.validity())?.into_nullable();
-
-        Ok(BoolArray::try_new(match_mask, validity)?.into_array())
+impl MaybeCompareFn for PrimitiveArray {
+    fn maybe_compare(&self, other: &Array, operator: Operator) -> Option<VortexResult<Array>> {
+        let fn_impl = || -> VortexResult<Array> {
```

Comment:
You immediately invoke this function - why make it a function?

Reply:
So I can have nice `?` for errors, but I'll just split it into a different function

Reply:
done